File: agenkit-ts/src/evaluation/bayesian-optimizer.ts
/**
 * Bayesian Optimization for Hyperparameter Tuning
 *
 * This module implements Bayesian optimization using a simplified surrogate model
 * based on local statistics. It balances exploration and exploitation through
 * acquisition functions (Expected Improvement, Upper Confidence Bound).
 *
 * Example:
 * ```typescript
 * const searchSpace = new SearchSpace();
 * searchSpace.addContinuous('temperature', 0.0, 1.0);
 * searchSpace.addContinuous('top_p', 0.0, 1.0);
 *
 * const optimizer = new BayesianOptimizer({
 *   agentFactory: (config) => new MyAgent(config),
 *   searchSpace,
 *   objective: 'accuracy',
 *   nInitial: 5,
 *   acquisitionFunction: 'ei',
 * });
 *
 * const result = await optimizer.optimize(testCases, 50);
 * console.log(`Best config: ${JSON.stringify(result.bestConfig)}`);
 * console.log(`Best score: ${result.bestScore.toFixed(3)}`);
 * ```
 */

import { Optimizer, SearchSpace, OptimizationResult, AgentFactory, ObjectiveFunction } from './optimizer';
import { TestCase } from './core';

/**
 * Acquisition function types.
 */
export type AcquisitionFunction = 'ei' | 'ucb' | 'pi';

/**
 * Configuration for Bayesian Optimizer.
 */
export interface BayesianOptimizerConfig {
  /** Function that creates agent from config */
  agentFactory: AgentFactory;
  /** SearchSpace defining parameter space */
  searchSpace: SearchSpace;
  /** Metric name or custom objective function */
  objective: string | ObjectiveFunction;
  /** Whether to maximize (true) or minimize (false) objective */
  maximize?: boolean;
  /** Acquisition function to use (default: 'ei') */
  acquisitionFunction?: AcquisitionFunction;
  /** Number of initial random samples (default: 5) */
  nInitial?: number;
  /** Exploration parameter for EI and PI (default: 0.01) */
  xi?: number;
  /** Exploration parameter for UCB (default: 2.576) */
  kappa?: number;
}

/**
 * Bayesian Optimizer for intelligent hyperparameter search.
 *
 * Uses a simplified surrogate model based on local statistics to guide
 * the search process. Balances exploration (trying new areas) with
 * exploitation (refining known good areas) using acquisition functions.
 */
export class BayesianOptimizer extends Optimizer {
  private acquisitionFunction: AcquisitionFunction;
  private nInitial: number;
  private xi: number;
  private kappa: number;
  private bestConfig: Record<string, unknown> | null = null;
  private bestScore: number = -Infinity;

  /**
   * Create Bayesian optimizer.
   *
   * @param config Optimizer configuration
   */
  constructor(config: BayesianOptimizerConfig) {
    super(
      config.agentFactory,
      config.searchSpace,
      config.objective,
      config.maximize ?? true
    );

    this.acquisitionFunction = config.acquisitionFunction || 'ei';
    this.nInitial = config.nInitial || 5;
    this.xi = config.xi || 0.01;
    this.kappa = config.kappa || 2.576; // 99% confidence interval
  }

  /**
   * Run Bayesian optimization.
   *
   * Algorithm:
   * 1. Sample n_initial random configurations
   * 2. Evaluate and build local statistics
   * 3. Use acquisition function to select next config
   * 4. Evaluate new config
   * 5. Update statistics and repeat
   *
   * @param testCases Test cases for evaluation
   * @param nIterations Number of iterations to run
   * @returns OptimizationResult with best configuration and history
   */
  async optimize(testCases: TestCase[], nIterations: number): Promise<OptimizationResult> {
    const startTime = new Date();
    this.history = [];
    this.bestConfig = null;
    this.bestScore = -Infinity;

    // Phase 1: Initial random exploration
    const nInitialEvals = Math.min(this.nInitial, nIterations);
    for (let i = 0; i < nInitialEvals; i++) {
      const config = this.searchSpace.sample();
      const score = await this.evaluateConfig(config, testCases);
      this.history.push([{ ...config }, score]);

      if (score > this.bestScore) {
        this.bestScore = score;
        this.bestConfig = { ...config };
      }
    }

    // Phase 2: Bayesian optimization
    for (let i = nInitialEvals; i < nIterations; i++) {
      // Select next configuration using acquisition function
      const config = await this.selectNextConfig();

      // Evaluate
      const score = await this.evaluateConfig(config, testCases);
      this.history.push([{ ...config }, score]);

      // Update best
      if (score > this.bestScore) {
        this.bestScore = score;
        this.bestConfig = { ...config };
      }
    }

    const endTime = new Date();

    // Fallback if no valid config found
    if (this.bestConfig === null) {
      this.bestConfig = this.searchSpace.sample();
    }

    return {
      bestConfig: this.bestConfig,
      bestScore: this.bestScore,
      history: this.history,
      nIterations,
      startTime,
      endTime,
      metadata: {
        algorithm: 'bayesian_optimization',
        acquisition_function: this.acquisitionFunction,
        n_initial: this.nInitial,
      },
    };
  }

  /**
   * Select next configuration to evaluate using acquisition function.
   *
   * Samples multiple candidates and selects the one with highest
   * acquisition value.
   *
   * @returns Next configuration to evaluate
   */
  private async selectNextConfig(): Promise<Record<string, unknown>> {
    // Sample multiple candidates
    const nCandidates = 100;
    const candidates: Record<string, unknown>[] = [];
    for (let i = 0; i < nCandidates; i++) {
      candidates.push(this.searchSpace.sample());
    }

    // Evaluate acquisition function for each candidate
    let bestCandidate = candidates[0];
    let bestAcquisition = -Infinity;

    for (const candidate of candidates) {
      const acquisition = this.computeAcquisition(candidate);
      if (acquisition > bestAcquisition) {
        bestAcquisition = acquisition;
        bestCandidate = candidate;
      }
    }

    return bestCandidate;
  }

  /**
   * Compute acquisition function value for a configuration.
   *
   * @param config Configuration to evaluate
   * @returns Acquisition value (higher = more promising)
   */
  private computeAcquisition(config: Record<string, unknown>): number {
    // Calculate local statistics (mean and std) from nearby points
    const { mean, std } = this.computeLocalStatistics(config);

    switch (this.acquisitionFunction) {
      case 'ei':
        return this.expectedImprovement(mean, std);
      case 'ucb':
        return this.upperConfidenceBound(mean, std);
      case 'pi':
        return this.probabilityOfImprovement(mean, std);
      default:
        return this.expectedImprovement(mean, std);
    }
  }

  /**
   * Compute local statistics (mean and std) from nearby evaluated points.
   *
   * Uses a simplified approach: weighted average based on distance in
   * parameter space.
   *
   * @param config Configuration to compute statistics for
   * @returns Mean and standard deviation estimates
   */
  private computeLocalStatistics(config: Record<string, unknown>): { mean: number; std: number } {
    if (this.history.length === 0) {
      return { mean: 0, std: 1 };
    }

    // Calculate distances to all evaluated points
    const distances: number[] = [];
    const scores: number[] = [];

    for (const [evalConfig, score] of this.history) {
      const distance = this.configDistance(config, evalConfig);
      distances.push(distance);
      scores.push(score);
    }

    // Use k-nearest neighbors (k=5) for local statistics
    const k = Math.min(5, this.history.length);
    const indices = this.argsort(distances).slice(0, k);

    // Calculate weighted mean and std
    let sumWeights = 0;
    let weightedSum = 0;

    for (const idx of indices) {
      // Weight decreases with distance (inverse distance weighting)
      const weight = 1.0 / (distances[idx] + 1e-10);
      weightedSum += scores[idx] * weight;
      sumWeights += weight;
    }

    const mean = weightedSum / sumWeights;

    // Calculate standard deviation
    let sumSquaredDiff = 0;
    for (const idx of indices) {
      const weight = 1.0 / (distances[idx] + 1e-10);
      sumSquaredDiff += weight * Math.pow(scores[idx] - mean, 2);
    }
    const std = Math.sqrt(sumSquaredDiff / sumWeights);

    return { mean, std: Math.max(std, 0.01) }; // Minimum std to avoid division by zero
  }

  /**
   * Calculate distance between two configurations.
   *
   * For continuous parameters: normalized Euclidean distance
   * For discrete/categorical: 0 if same, 1 if different
   *
   * @param config1 First configuration
   * @param config2 Second configuration
   * @returns Distance between configurations
   */
  private configDistance(
    config1: Record<string, unknown>,
    config2: Record<string, unknown>
  ): number {
    let distance = 0;
    let count = 0;

    for (const name of this.searchSpace.getParameterNames()) {
      const param = this.searchSpace.getParameter(name);
      if (!param) continue;

      const val1 = config1[name];
      const val2 = config2[name];

      if (param.type === 'continuous' || param.type === 'integer') {
        // Normalized Euclidean distance
        const range = param.high! - param.low!;
        const normalizedDiff = (Number(val1) - Number(val2)) / range;
        distance += normalizedDiff * normalizedDiff;
        count++;
      } else {
        // Discrete/categorical: binary distance
        distance += val1 === val2 ? 0 : 1;
        count++;
      }
    }

    return count > 0 ? Math.sqrt(distance / count) : 0;
  }

  /**
   * Expected Improvement acquisition function.
   *
   * @param mean Predicted mean
   * @param std Predicted standard deviation
   * @returns EI value
   */
  private expectedImprovement(mean: number, std: number): number {
    if (std < 1e-10) {
      return 0;
    }

    const improvement = mean - this.bestScore - this.xi;
    const z = improvement / std;

    // EI = improvement * Φ(z) + std * φ(z)
    // where Φ is CDF and φ is PDF of standard normal
    const ei = improvement * this.normalCdf(z) + std * this.normalPdf(z);

    return Math.max(ei, 0);
  }

  /**
   * Upper Confidence Bound acquisition function.
   *
   * @param mean Predicted mean
   * @param std Predicted standard deviation
   * @returns UCB value
   */
  private upperConfidenceBound(mean: number, std: number): number {
    return mean + this.kappa * std;
  }

  /**
   * Probability of Improvement acquisition function.
   *
   * @param mean Predicted mean
   * @param std Predicted standard deviation
   * @returns PI value
   */
  private probabilityOfImprovement(mean: number, std: number): number {
    if (std < 1e-10) {
      return 0;
    }

    const improvement = mean - this.bestScore - this.xi;
    const z = improvement / std;

    return this.normalCdf(z);
  }

  /**
   * Standard normal cumulative distribution function (CDF).
   *
   * @param x Input value
   * @returns Φ(x)
   */
  private normalCdf(x: number): number {
    // Using error function approximation
    return 0.5 * (1 + this.erf(x / Math.sqrt(2)));
  }

  /**
   * Standard normal probability density function (PDF).
   *
   * @param x Input value
   * @returns φ(x)
   */
  private normalPdf(x: number): number {
    return Math.exp(-0.5 * x * x) / Math.sqrt(2 * Math.PI);
  }

  /**
   * Error function approximation.
   *
   * @param x Input value
   * @returns erf(x)
   */
  private erf(x: number): number {
    // Abramowitz and Stegun approximation
    const sign = x >= 0 ? 1 : -1;
    x = Math.abs(x);

    const a1 = 0.254829592;
    const a2 = -0.284496736;
    const a3 = 1.421413741;
    const a4 = -1.453152027;
    const a5 = 1.061405429;
    const p = 0.3275911;

    const t = 1.0 / (1.0 + p * x);
    const y = 1.0 - ((((a5 * t + a4) * t + a3) * t + a2) * t + a1) * t * Math.exp(-x * x);

    return sign * y;
  }

  /**
   * Get indices that would sort an array.
   *
   * @param arr Array to sort
   * @returns Sorted indices
   */
  private argsort(arr: number[]): number[] {
    return arr
      .map((val, idx) => ({ val, idx }))
      .sort((a, b) => a.val - b.val)
      .map(item => item.idx);
  }
}

File: agenkit/evaluation/bayesian_optimizer.py
"""
Bayesian Optimization for Hyperparameter Tuning

Uses Gaussian Process regression to build a surrogate model of the
objective function and selects next configurations using acquisition
functions (Expected Improvement, Upper Confidence Bound).

Example:
    >>> from agenkit.evaluation import BayesianOptimizer
    >>>
    >>> optimizer = BayesianOptimizer(
    ...     agent_factory=lambda config: MyAgent(**config),
    ...     search_space={"temperature": (0.0, 1.0), "top_p": (0.0, 1.0)},
    ...     objective="accuracy",
    ...     acquisition="ei"
    ... )
    >>>
    >>> result = await optimizer.optimize(test_cases, n_iterations=50)
"""

from collections.abc import Callable
from datetime import datetime, timezone
from enum import Enum
from typing import Any

import numpy as np
from sklearn.gaussian_process import GaussianProcessRegressor
from sklearn.gaussian_process.kernels import Matern

from .optimizer import OptimizationResult, Optimizer, SearchSpace


class AcquisitionFunction(Enum):
    """Acquisition function types for Bayesian optimization."""

    EI = "ei"  # Expected Improvement
    UCB = "ucb"  # Upper Confidence Bound
    PI = "pi"  # Probability of Improvement


class BayesianOptimizer(Optimizer):
    """
    Bayesian optimization using Gaussian Process.

    Uses a probabilistic surrogate model to intelligently select
    configurations to evaluate, balancing exploration and exploitation.

    Algorithm:
        1. Sample n_initial random configurations
        2. Evaluate and fit Gaussian Process
        3. Use acquisition function to select next config
        4. Evaluate new config
        5. Update GP and repeat

    Example:
        >>> optimizer = BayesianOptimizer(
        ...     agent_factory=lambda c: MyAgent(**c),
        ...     search_space={"temperature": (0.0, 1.0)},
        ...     objective="accuracy",
        ...     acquisition="ei",
        ...     n_initial=10
        ... )
        >>> result = await optimizer.optimize(test_cases, n_iterations=50)
    """

    def __init__(
        self,
        agent_factory: Callable[[dict[str, Any]], Any],
        search_space: SearchSpace | dict[str, Any],
        objective: str | Callable,
        maximize: bool = True,
        acquisition: str = "ei",
        n_initial: int = 5,
        xi: float = 0.01,
        kappa: float = 2.576,
        kernel: Any | None = None,
    ):
        """
        Initialize Bayesian optimizer.

        Args:
            agent_factory: Function that creates agent from config
            search_space: SearchSpace or dict defining parameter space
            objective: Metric name or custom objective function
            maximize: Whether to maximize (True) or minimize (False)
            acquisition: Acquisition function ("ei", "ucb", "pi")
            n_initial: Number of random initial samples
            xi: Exploration parameter for EI and PI (default: 0.01)
            kappa: Exploration parameter for UCB (default: 2.576 for 99% confidence)
            kernel: Gaussian Process kernel (default: Matern)
        """
        super().__init__(agent_factory, search_space, objective, maximize)

        self.acquisition_func = AcquisitionFunction(acquisition)
        self.n_initial = n_initial
        self.xi = xi
        self.kappa = kappa

        # Initialize Gaussian Process
        if kernel is None:
            kernel = Matern(nu=2.5)
        self.gp = GaussianProcessRegressor(kernel=kernel, n_restarts_optimizer=5, random_state=42)

        # Track parameter names and bounds for continuous parameters
        self.param_names: list[str] = []
        self.param_bounds: list[tuple] = []
        self._setup_continuous_space()

    def _setup_continuous_space(self) -> None:
        """Setup continuous parameter space for GP optimization."""
        for name, spec in self.search_space.parameters.items():
            if spec["type"] in ("continuous", "integer"):
                self.param_names.append(name)
                self.param_bounds.append((spec["low"], spec["high"]))

    def _config_to_vector(self, config: dict[str, Any]) -> np.ndarray:
        """Convert configuration dict to vector for GP."""
        return np.array([config[name] for name in self.param_names])

    def _vector_to_config(self, vector: np.ndarray) -> dict[str, Any]:
        """Convert vector to configuration dict."""
        config = {}
        for i, name in enumerate(self.param_names):
            spec = self.search_space.parameters[name]
            if spec["type"] == "integer":
                config[name] = round(vector[i])
            else:
                config[name] = float(vector[i])

        # Add categorical/discrete parameters (sample randomly for now)
        for name, spec in self.search_space.parameters.items():
            if name not in config and (spec["type"] == "discrete" or spec["type"] == "categorical"):
                config[name] = np.random.choice(spec["values"])

        return config

    def _expected_improvement(
        self,
        X: np.ndarray,  # noqa: N803
        X_sample: np.ndarray,  # noqa: N803
        Y_sample: np.ndarray,  # noqa: N803
    ) -> np.ndarray:
        """
        Expected Improvement acquisition function.

        Args:
            X: Points to evaluate
            X_sample: Observed points
            Y_sample: Observed values

        Returns:
            Expected improvement at each point in X
        """
        mu, sigma = self.gp.predict(X, return_std=True)
        mu_sample_opt = np.max(Y_sample)

        with np.errstate(divide="warn"):
            imp = mu - mu_sample_opt - self.xi
            Z = imp / sigma  # noqa: N806
            ei = imp * self._norm_cdf(Z) + sigma * self._norm_pdf(Z)
            ei[sigma == 0.0] = 0.0

        return ei

    def _upper_confidence_bound(self, X: np.ndarray) -> np.ndarray:  # noqa: N803
        """
        Upper Confidence Bound acquisition function.

        Args:
            X: Points to evaluate

        Returns:
            UCB values at each point in X
        """
        mu, sigma = self.gp.predict(X, return_std=True)
        return mu + self.kappa * sigma

    def _probability_of_improvement(
        self,
        X: np.ndarray,  # noqa: N803
        X_sample: np.ndarray,  # noqa: N803
        Y_sample: np.ndarray,  # noqa: N803
    ) -> np.ndarray:
        """
        Probability of Improvement acquisition function.

        Args:
            X: Points to evaluate
            X_sample: Observed points
            Y_sample: Observed values

        Returns:
            Probability of improvement at each point in X
        """
        mu, sigma = self.gp.predict(X, return_std=True)
        mu_sample_opt = np.max(Y_sample)

        with np.errstate(divide="warn"):
            Z = (mu - mu_sample_opt - self.xi) / sigma  # noqa: N806
            pi = self._norm_cdf(Z)
            pi[sigma == 0.0] = 0.0

        return pi

    @staticmethod
    def _norm_cdf(x: np.ndarray) -> np.ndarray:
        """Standard normal cumulative distribution function."""
        import math

        return 0.5 * (1.0 + np.vectorize(lambda z: math.erf(z / math.sqrt(2.0)))(x))

    @staticmethod
    def _norm_pdf(x: np.ndarray) -> np.ndarray:
        """Standard normal probability density function."""
        return np.exp(-0.5 * x**2) / np.sqrt(2.0 * np.pi)

    def _propose_location(
        self,
        X_sample: np.ndarray,  # noqa: N803
        Y_sample: np.ndarray,  # noqa: N803
        n_candidates: int = 1000,
    ) -> np.ndarray:
        """
        Propose next location to sample using acquisition function.

        Args:
            X_sample: Observed points
            Y_sample: Observed values
            n_candidates: Number of random candidates to evaluate

        Returns:
            Next point to sample
        """
        # Generate random candidates
        candidates = np.random.uniform(
            low=[b[0] for b in self.param_bounds],
            high=[b[1] for b in self.param_bounds],
            size=(n_candidates, len(self.param_bounds)),
        )

        # Evaluate acquisition function
        if self.acquisition_func == AcquisitionFunction.EI:
            acq_values = self._expected_improvement(candidates, X_sample, Y_sample)
        elif self.acquisition_func == AcquisitionFunction.UCB:
            acq_values = self._upper_confidence_bound(candidates)
        elif self.acquisition_func == AcquisitionFunction.PI:
            acq_values = self._probability_of_improvement(candidates, X_sample, Y_sample)
        else:
            msg = f"Unknown acquisition function: {self.acquisition_func}"
            raise ValueError(msg)

        # Return candidate with highest acquisition value
        best_idx = np.argmax(acq_values)
        return candidates[best_idx]

    async def optimize(
        self, test_cases: list[dict[str, Any]], n_iterations: int, **kwargs: Any
    ) -> OptimizationResult:
        """
        Run Bayesian optimization.

        Args:
            test_cases: Test cases for evaluation
            n_iterations: Total number of iterations (including initial samples)
            **kwargs: Additional parameters

        Returns:
            OptimizationResult with best configuration
        """
        start_time = datetime.now(timezone.utc).isoformat()
        self.history = []

        # Phase 1: Random initialization
        X_sample: list[np.ndarray] = []  # noqa: N806
        Y_sample: list[float] = []  # noqa: N806

        for _ in range(min(self.n_initial, n_iterations)):
            config = self.search_space.sample()
            score = await self.evaluate_config(config, test_cases)

            self.history.append((config.copy(), score))
            X_sample.append(self._config_to_vector(config))
            Y_sample.append(score)

        # Phase 2: Bayesian optimization
        for _ in range(self.n_initial, n_iterations):
            # Fit GP on observed data
            X_array = np.array(X_sample)  # noqa: N806
            Y_array = np.array(Y_sample)  # noqa: N806
            self.gp.fit(X_array, Y_array)

            # Propose next location
            next_x = self._propose_location(X_array, Y_array)
            next_config = self._vector_to_config(next_x)

            # Evaluate
            score = await self.evaluate_config(next_config, test_cases)

            # Update history
            self.history.append((next_config.copy(), score))
            X_sample.append(next_x)
            Y_sample.append(score)

        # Find best configuration
        best_idx = int(np.argmax(Y_sample))
        best_config = self.history[best_idx][0]
        best_score = Y_sample[best_idx]

        end_time = datetime.now(timezone.utc).isoformat()

        return OptimizationResult(
            best_config=best_config,
            best_score=best_score,
            history=self.history,
            n_iterations=n_iterations,
            start_time=start_time,
            end_time=end_time,
            metadata={
                "algorithm": "bayesian_optimization",
                "acquisition": self.acquisition_func.value,
                "n_initial": self.n_initial,
                "kernel": str(self.gp.kernel_),
            },
        )
